FastBuild Generator: some fixes for custom command execution

packadal opened the pull request:
- properly escape single quotes in strings
- re-ordered variables and pre-build  dependencies

Commit:
FastBuild Generator: some fixes for custom command execution

* properly escape single quotes in strings
* re-ordered variables and pre-build  dependencies

packadal commented:
Is this more what you had in mind ?

Commit:
unescape variables that were escaped with '^'

Commit:
fixed relative path for MSVC 12 2013

Commit:
do not write alias for empty config groups

inbilla commented on Source/cmGlobalFastbuildGenerator.cxx:
```diff
@@ -2202,10 +2219,10 @@ class cmGlobalFastbuildGenerator::Detail::Generation
 					context.fc.WriteArray("PreBuildDependencies",
 						Wrap(preBuildSteps, "'" + targetName + "-", "-" + configName + "'"),
-						"+");
+						"=");
```
I think this should still be a "+" (and there should be a using statement above, see other comment)

inbilla commented on Source/cmGlobalFastbuildGenerator.cxx:
```diff
@@ -2185,6 +2184,24 @@ class cmGlobalFastbuildGenerator::Detail::Generation
+
+			context.fc.WriteVariable("BaseCompilationConfig_" + configName, "");
+			context.fc.WritePushScopeStruct();
+
```
here, there should be this:
context.fc.WriteCommand("Using", ".BaseConfig_" + configName);

inbilla commented:
added a couple of smaller comments, I think it's almost ready to go.

packadal commented:
OK, I fixed the last things you mentionned.
I also added a few more fixes on top of the branch, if you want I can move them to another branch.

inbilla commented:
I think that should be ok, have you had a chance to run unit tests?
If not, I'll try run them myself before merging

packadal commented:
Here's the full list of failing tests: 

80% tests passed, 73 tests failed out of 371

Label Time Summary:
Label1    =   0.82 sec
Label2    =   0.82 sec

Total Test time (real) = 859.98 sec

The following tests FAILED:
          1 - SystemInformationNew (Failed)
         44 - ObjectLibrary (Failed)
         50 - LinkDirectory (Failed)
         58 - SourceGroups (Failed)
         59 - Preprocess (Failed)
         60 - ExportImport (Failed)
         67 - AliasTarget (Failed)
         68 - StagingPrefix (Failed)
         70 - ConfigSources (Failed)
         74 - BundleUtilities (Failed)
         78 - Module.ExternalData (Failed)
         79 - Module.GenerateExportHeader (Failed)
        100 - SubProject (Failed)
        101 - SubProject-Stage2 (Failed)
        108 - CustomCommand (Failed)
        109 - CustomCommandByproducts (Failed)
        111 - CustomCommandWorkingDirectory (Failed)
        112 - OutOfSource (Failed)
        113 - BuildDepends (Failed)
        114 - SimpleInstall (Failed)
        115 - SimpleInstall-Stage2 (Failed)
        127 - complex (Failed)
        128 - complexOneConfig (Failed)
        131 - ExternalProject (Failed)
        132 - ExternalProjectLocal (Failed)
        133 - ExternalProjectUpdateSetup (Failed)
        134 - ExternalProjectUpdate (Failed)
        149 - JumpWithLibOut (Failed)
        150 - JumpNoLibOut (Failed)
        151 - Plugin (Failed)
        154 - SubDir (Failed)
        156 - PDBDirectoryAndName (Failed)
        157 - PrecompiledHeader (Failed)
        158 - ModuleDefinition (Failed)
        160 - MFC (Failed)
        184 - CTestConfig.Script.Debug (Failed)
        185 - CTestConfig.Dashboard.Debug (Failed)
        186 - CTestConfig.Script.MinSizeRel (Failed)
        187 - CTestConfig.Dashboard.MinSizeRel (Failed)
        188 - CTestConfig.Script.Release (Failed)
        189 - CTestConfig.Dashboard.Release (Failed)
        190 - CTestConfig.Script.RelWithDebInfo (Failed)
        191 - CTestConfig.Dashboard.RelWithDebInfo (Failed)
        197 - CMakeCommands.target_compile_options (Failed)
        214 - CTestTestTimeout (Failed)
        224 - IncludeDirectories (Failed)
        233 - CMakeOnly.LinkInterfaceLoop (Failed)
        234 - CMakeOnly.CheckSymbolExists (Failed)
        235 - CMakeOnly.CheckCXXSymbolExists (Failed)
        236 - CMakeOnly.CheckCXXCompilerFlag (Failed)
        237 - CMakeOnly.CheckLanguage (Failed)
        238 - CMakeOnly.CheckStructHasMember (Failed)
        239 - CMakeOnly.CompilerIdC (Failed)
        240 - CMakeOnly.CompilerIdCXX (Failed)
        241 - CMakeOnly.AllFindModules (Failed)
        242 - CMakeOnly.SelectLibraryConfigurations (Failed)
        243 - CMakeOnly.TargetScope (Failed)
        244 - CMakeOnly.find_library (Failed)
        245 - CMakeOnly.find_path (Failed)
        246 - CMakeOnly.ProjectInclude (Failed)
        247 - CMakeOnly.MajorVersionSelection-PythonLibs_2 (Failed)
        248 - CMakeOnly.MajorVersionSelection-PythonLibs_3 (Failed)
        249 - CMakeOnly.MajorVersionSelection-PythonInterp_2 (Failed)
        250 - CMakeOnly.MajorVersionSelection-PythonInterp_3 (Failed)
        251 - CMakeOnly.MajorVersionSelection-Qt_3 (Failed)
        252 - CMakeOnly.MajorVersionSelection-Qt_4 (Failed)
        273 - RunCMake.ctest_memcheck (Failed)
        275 - RunCMake.Configure (Failed)
        300 - RunCMake.ctest_build (Failed)
        301 - RunCMake.ctest_configure (Failed)
        304 - RunCMake.ctest_test (Failed)
        331 - RunCMake.File_Generate (Failed)
        371 - CMake.CheckSourceTree (Failed)

packadal commented:
this looks strange, let me check it was a clean build of this branch an make a second run

packadal commented:
ah, fbuild.exe was not in the path on this run. 
Here are the actual test results (sorry for the noise)

89% tests passed, 40 tests failed out of 371

Label Time Summary:
Label1    =   0.24 sec
Label2    =   0.24 sec

Total Test time (real) = 710.71 sec

The following tests FAILED:
         58 - SourceGroups (Failed)
         59 - Preprocess (Failed)
         60 - ExportImport (Failed)
         68 - StagingPrefix (Failed)
         70 - ConfigSources (Failed)
         78 - Module.ExternalData (Failed)
         79 - Module.GenerateExportHeader (Failed)
        100 - SubProject (Failed)
        101 - SubProject-Stage2 (Failed)
        107 - GeneratorExpression (Failed)
        108 - CustomCommand (Failed)
        109 - CustomCommandByproducts (Failed)
        112 - OutOfSource (Failed)
        113 - BuildDepends (Failed)
        114 - SimpleInstall (Failed)
        115 - SimpleInstall-Stage2 (Failed)
        127 - complex (Failed)
        128 - complexOneConfig (Failed)
        131 - ExternalProject (Failed)
        132 - ExternalProjectLocal (Failed)
        133 - ExternalProjectUpdateSetup (Failed)
        134 - ExternalProjectUpdate (Failed)
        151 - Plugin (Failed)
        154 - SubDir (Failed)
        157 - PrecompiledHeader (Failed)
        158 - ModuleDefinition (Failed)
        160 - MFC (Failed)
        184 - CTestConfig.Script.Debug (Failed)
        185 - CTestConfig.Dashboard.Debug (Failed)
        186 - CTestConfig.Script.MinSizeRel (Failed)
        187 - CTestConfig.Dashboard.MinSizeRel (Failed)
        188 - CTestConfig.Script.Release (Failed)
        189 - CTestConfig.Dashboard.Release (Failed)
        190 - CTestConfig.Script.RelWithDebInfo (Failed)
        191 - CTestConfig.Dashboard.RelWithDebInfo (Failed)
        197 - CMakeCommands.target_compile_options (Failed)
        224 - IncludeDirectories (Failed)
        238 - CMakeOnly.CheckStructHasMember (Failed)
        275 - RunCMake.Configure (Failed)
        331 - RunCMake.File_Generate (Failed)
Errors while running CTest

Commit:
output commands into a script file that is then run by Fastbuild

This is necessary as fastbuild uses windows' CreateProcess, which does not support output redirections with '>'.
This may work on linux, but I cannot test it.

The test 'CustomCommand' finally passes

Removed an unused function and an unused functor struct

packadal commented:
I added a new commit that moves the commands into a .bat file, and now the 'CustomCommand' test passes !

here is a full log of the test results:

90% tests passed, 38 tests failed out of 371

Label Time Summary:
Label1    =   0.05 sec
Label2    =   0.05 sec

Total Test time (real) = 770.54 sec

The following tests FAILED:
         58 - SourceGroups (Failed)
         59 - Preprocess (Failed)
         60 - ExportImport (Failed)
         68 - StagingPrefix (Failed)
         70 - ConfigSources (Failed)
         79 - Module.GenerateExportHeader (Failed)
        100 - SubProject (Failed)
        101 - SubProject-Stage2 (Failed)
        109 - CustomCommandByproducts (Failed)
        112 - OutOfSource (Failed)
        113 - BuildDepends (Failed)
        114 - SimpleInstall (Failed)
        115 - SimpleInstall-Stage2 (Failed)
        127 - complex (Failed)
        128 - complexOneConfig (Failed)
        131 - ExternalProject (Failed)
        132 - ExternalProjectLocal (Failed)
        133 - ExternalProjectUpdateSetup (Failed)
        134 - ExternalProjectUpdate (Failed)
        151 - Plugin (Failed)
        154 - SubDir (Failed)
        157 - PrecompiledHeader (Failed)
        158 - ModuleDefinition (Failed)
        160 - MFC (Failed)
        184 - CTestConfig.Script.Debug (Failed)
        185 - CTestConfig.Dashboard.Debug (Failed)
        186 - CTestConfig.Script.MinSizeRel (Failed)
        187 - CTestConfig.Dashboard.MinSizeRel (Failed)
        188 - CTestConfig.Script.Release (Failed)
        189 - CTestConfig.Dashboard.Release (Failed)
        190 - CTestConfig.Script.RelWithDebInfo (Failed)
        191 - CTestConfig.Dashboard.RelWithDebInfo (Failed)
        197 - CMakeCommands.target_compile_options (Failed)
        224 - IncludeDirectories (Failed)
        238 - CMakeOnly.CheckStructHasMember (Failed)
        275 - RunCMake.Configure (Failed)
        331 - RunCMake.File_Generate (Failed)
        371 - CMake.CheckSourceTree (Failed)
Errors while running CTest

inbilla commented:
Interesting, on my current configuration (without this merged in) I'm currently getting 
100 - SubProject and
371 - CMake.CheckSourceTree as a PASS

But I also get a few others failing.
371 appears to be a new failure (compared to your previous test run too)

This is a good improvement :)
I'm keen to get this in, I'm not sure about test 371 and some of the details...
I think we can iterate on this though
